model/src/component/button.rs

```diff
@@ -0,0 +1,144 @@
+    pub id: Option<EmojiId>,
+    pub name: String,
+    pub animated: bool,
+}
+
+/// Helper struct for assembling  the Button struct
+#[derive(Debug)]
+pub struct ButtonBuilder {
+    pub(crate) style: ButtonStyle,
+    pub(crate) emoji: Option<PartialEmoji>,
+    pub(crate) label: Option<String>,
+    pub(crate) custom_id: Option<String>,
+    pub(crate) url: Option<String>,
+    pub(crate) disabled: bool,
+}
+
+#[derive(Debug)]
+pub enum ButtonBuilderError {
+    MissingLabelOrEmoji,
+    LabelTooLong,
+    CustomIdTooLong,
+    LinkTooLong,
+}
+
+impl ButtonBuilder {
+    /// Makes a new button that links to an external site
+    /// You will also need to set the emoji or label as well (or both)
+    ///
+    /// # Errors
+    /// Emits `ButtonBuilderError::LinkTooLong` if the link is over 512 chars
+    pub fn new_link(url: impl Into<String>) -> Result<Self, ButtonBuilderError> {
+        let url = url.into();
+        if url.len() > 512 {
+            Err(ButtonBuilderError::LinkTooLong)
+        } else {
+            Ok(Self {
+                style: ButtonStyle::Link,
+                emoji: None,
+                label: None,
+                custom_id: None,
+                url: Some(url),
+                disabled: false,
+            })
+        }
+    }
+
+    /// Make a new button with a specific style and id
+    /// You will also need to set the emoji or label as well (or both)
+    ///
+    /// # Errors
+    /// Emits `ButtonBuilderError::CustomIdTooLong` if  the custom id is over 100 chars
+    pub fn new_button(
+        style: ButtonStyle,
+        custom_id: impl Into<String>,
+    ) -> Result<Self, ButtonBuilderError> {
+        let custom_id = custom_id.into();
+        if custom_id.len() > 100 {
+            Err(ButtonBuilderError::CustomIdTooLong)
+        } else {
+            Ok(Self {
+                style,
+                emoji: None,
+                label: None,
+                custom_id: Some(custom_id),
+                url: None,
+                disabled: false,
+            })
+        }
+    }
+
+    /// Set an emoji for the button
+    pub fn with_emoji(mut self, emoji: PartialEmoji) -> Self {
+        self.emoji = Some(emoji);
+        self
+    }
+
+    /// Set the button label text
+    ///
+    /// # Errors
+    /// Emits `ButtonBuilderError::LabelTooLong` if the label is more than 80 characters
+    pub fn with_label(mut self, label: impl Into<String>) -> Result<Self, ButtonBuilderError> {
+        let label = label.into();
+        if label.len() > 80 {
+            Err(ButtonBuilderError::LabelTooLong)
+        } else {
+            self.label = Some(label);
+            Ok(self)
+        }
+    }
+
+    /// Disable the button
+    pub fn disabled(mut self) -> Self {
+        self.disabled = true;
+        self
+    }
+
+    /// Finish the builder and assemble into the component
+    ///
+    /// # Errors
+    /// Emits `ButtonBuilderError::MissingLabelOrEmoji` when there is no label or emoji set
```

Comment:
Please format these docs similarly to other docs: One line for a short description, followed by lines about other things. There needs to be a newline after the error section. The Error type needs to link to the error itself.